From 8dcf282bcec842ae633f43fc6dd1ecb397986d5c Mon Sep 17 00:00:00 2001
From: oshmoun <oshmoun100@googlemail.com>
Date: Fri, 10 Aug 2018 09:28:52 +0200
Subject: [PATCH] msm8998: gralloc1: disable UBWC if video encoder client has
 no support

On msm8956, vidc does not support UBWC, but the display hal is setting the flag regardless, because adreno is reporting support.
This leads to issues when trying to encode the actual display of the device, i.e when screen recording or when casting.
Detect such cases by checking if:
1. The client is video encoder
2. It has declared inability to handle UBWC through a property

Test: builds ok - manual testing.

Change-Id: I1ff2489b0ce8fe36a801881b848873e591077402
---
 msm8998/libgralloc1/gr_allocator.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/msm8998/libgralloc1/gr_allocator.cpp b/msm8998/libgralloc1/gr_allocator.cpp
index f82cb6ce..a6886e01 100644
--- a/msm8998/libgralloc1/gr_allocator.cpp
+++ b/msm8998/libgralloc1/gr_allocator.cpp
@@ -602,6 +602,9 @@ void Allocator::GetIonHeapInfo(gralloc1_producer_usage_t prod_usage,
 
 bool Allocator::IsUBwcEnabled(int format, gralloc1_producer_usage_t prod_usage,
                               gralloc1_consumer_usage_t cons_usage) {
+  // Property is used to check whether the video encoder supports UBWC
+  char property[PROPERTY_VALUE_MAX];
+
   // Allow UBWC, if client is using an explicitly defined UBWC pixel format.
   if (IsUBwcFormat(format)) {
     return true;
@@ -623,6 +626,12 @@ bool Allocator::IsUBwcEnabled(int format, gralloc1_producer_usage_t prod_usage,
       enable = adreno_helper_->IsUBWCSupportedByGPU(format);
     }
 
+    // Check if client is video encoder, and UBWC is disabled by a prop
+    if ((cons_usage & GRALLOC1_CONSUMER_USAGE_VIDEO_ENCODER) &&
+        (property_get("video.disable.ubwc", property, "0") > 0)) {
+      enable = atoi(property) == 0;
+    }
+
     // Allow UBWC, only if CPU usage flags are not set
     if (enable && !(CpuCanAccess(prod_usage, cons_usage))) {
       return true;
-- 
2.15.1

